### ISSUE TITLE
IVsProjectRestoreInfoSource.WhenNominated returns immediately even when a nomination is pending

### ISSUE THREAD

**Issue by nkolev92** [dotnet/project-system#7453]
**Visual Studio Version**:
17.0 preview 3

**Summary**:

IVsProjectRestoreInfoSource.WhenNominated returns immediately even when `HasPendingNomination` is true. 

This basically causes NuGet to have to invoke `WhenNominated` multiple times on the same project.

This basically leads to NuGet not being able to accurately reason about the effectiveness of the batching, as it pollutes the data with a high number of checks. 

NuGet still gets to batch the restores, but there are caveats. 

* What is currently happening is effectively busy spin. A threadpool is blocking doing a large amount of work.
* It's technically still more effective than what was happening before, but we're pay more than we need to.

To fully understand what's going on, I added the following logging: https://github.com/NuGet/NuGet.Client/compare/dev-nkolev92-logging?expand=1. 

I have added the logs at the bottom.

**Steps to Reproduce**:

1. Clone https://github.com/NuGet/NuGet.Client
1. Checkout `dev-nkolev92-testBranchSwitch` on the commandline.
1. Load NuGet-Src-Commandline.slnf and observe 1 restore.
1. Branch switch to `dev-nkolev92-testBranchSwitch2` in Visual Studio and observe 1 restore.
1. The branch differences can be observed at https://github.com/NuGet/NuGet.Client/compare/dev-nkolev92-testBranchSwitch2...dev-nkolev92-testBranchSwitch

You can observe the vs/nuget/restoreinformation telemetry and see that ProjectsReadyCheckCount is 100+ instead of the expected 2.

**Expected Behavior**:

WhenNominated returns only after the project design time build has completed.

**Actual Behavior**:

WhenNominated returns immediately. 

**User Impact**:

* Inability to reason about the success of the feature easily.
* Busy threadpool thread, slower batching.

cc @ocallesp @lifengl 
[RestoreCheckLogs.txt](https://github.com/dotnet/project-system/files/6934483/RestoreCheckLogs.txt)

**Comment by ocallesp** [comment]
Pending: check _whenNominatedTask

**Comment by nkolev92** [comment]
Closing in favor of https://github.com/NuGet/Home/issues/11132
